### PR TITLE
Feat/open eth app auto

### PR DESCRIPTION
**Pull request by fstream9**
**Description**
 Added funcationality to open Ethereum app if not already opened and logic to also close any other app that is currently open before opening Ethereum app.

**Checklist**

* [ ] There is a related GitHub issue
* [ ] Tests are included if applicable
* [ ] Any added code is fully documented

**Issue**

Progresses #???

### CODE REVIEW

**Review by GummyGod — APPROVED** [review]

File: app/components/Views/LedgerConnect/index.tsx [hunk 1]
```diff
@@ -65,46 +66,59 @@ const LedgerConnect = () => {
...
-				setIsRetry(true);
-				setIsConnecting(false);
+
+			if (appName === 'BOLOS') {
+				// Open Ethereum App
+				Alert.alert(strings('ledger.ethereum_app_open'), strings('ledger.ethereum_app_open_message'));
+				try {
+					await openLedgerEthApp(transportRef as React.MutableRefObject<BluetoothTransport>);
+				} catch (e) {
+					throw new Error(strings('ledger.ethereum_app_open_error'));
```
[the remark this answers is not in the document]
Comment (fstream9) [reply app/components/Views/LedgerConnect/index.tsx:111]: This error gets passed onto the outer catch which has the expected behaviour

File: app/components/Views/LedgerConnect/ledgerUtils.ts [hunk 2]
```diff
@@ -51,3 +52,9 @@ export const handleAndroidBluetoothPermissions = async (
 
 	return false;
 };
+
+export const openLedgerEthApp = async (transportRef: React.MutableRefObject<BluetoothTransport>): Promise<Buffer> =>
+	transportRef.current.send(0xe0, 0xd8, 0x00, 0x00, Buffer.from('Ethereum', 'ascii'));
```
[the remark this answers is not in the document]
Comment (fstream9) [reply app/components/Views/LedgerConnect/ledgerUtils.ts:57]: I agree

File: app/components/Views/LedgerConnect/index.tsx [hunk 3]
```diff
@@ -65,46 +66,59 @@ const LedgerConnect = () => {
...
+			try {
 				transportRef.current = await BluetoothTransport.open(selectedDevice);
-				transportRef.current?.on('disconnect', () => (transportRef.current = undefined));
-			}
-		} catch {
-			setIsRetry(true);
-			setIsConnecting(false);
+				transportRef.current?.on('disconnect', () => {
+					transportRef.current = undefined;
+					setHasDisconnected(true);
```
[the remark this answers is not in the document]
Comment (fstream9) [reply app/components/Views/LedgerConnect/index.tsx:85]: makes sense, I guess I didn't revisit the logic after I changed it from being state based to ref based

File: app/components/Views/LedgerConnect/index.tsx [hunk 4]
```diff
@@ -124,10 +138,27 @@ const LedgerConnect = () => {
...
 		}
 	};
 
+	useEffect(() => {
+		const checkLedgerConnection = async () => {
+			if (transportRef.current && !isRestartConnect) {
+				transportRef.current.close();
+			}
+
+			if (isConnecting && hasDisconnected && isRestartConnect.current) {
```
[the remark this answers is not in the document]
Comment (fstream9) [reply app/components/Views/LedgerConnect/index.tsx:153]: concerned there'll be a race condition from when isRestartConnect.current is set to true and the connection disconnecting and ensure we only do it when we meet the specific criteria for it to re-run. Ie, if it disconnects before we tell it it should be good to rerun, or has disconnected for some other reason then we should be sure it only reruns given the flag was set as it was an intentional restart

[the remark this answers is not in the document]
Comment (zoltanarvai) [reply app/components/Views/LedgerConnect/index.tsx:153]: Gotcha.

**Comment by zoltanarvai** [comment]
I deleted ledgerUtils.ts on the main branch.

File: app/components/Views/LedgerConnect/ledgerUtils.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (zoltanarvai) [reply app/components/Views/LedgerConnect/ledgerUtils.ts:57]: this has been moved to the controller, if you merge from the ledger-integrations branch you can remove the ledgerUtils and use KeyringController.openXYZ function (same for quit)

File: app/components/hooks/useLedgerBluetooth.tsx [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+		shouldRestartConnection: boolean;
+		restartCount: number;
+	}>({
+		shouldRestartConnection: false,
+		restartCount: 0,
+	});
+
+	// Due to the async nature of the disconnects after sending an APDU command we load the code to run in a stack
+	// with code being pushed and popped off of the stack
+	const logicToRun = useRef<(() => Promise<void>)[]>([]);
```
Comment (zoltanarvai) [inline app/components/hooks/useLedgerBluetooth.tsx:41]: What do you think of changing the name of this? Like processStack or workflowSteps or something. Currently is a bit colliding with the ledgerLogicToRun naming.

**Review by zoltanarvai — COMMENTED** [review]
Really sweet stuff

**Pull request merged**